### DISCUSSION TITLE
[Spring]JPQL 질문

### DISCUSSION THREAD

**Discussion by DoohyunHwang97** [wafflestudio/seminar-2023#165] [Q&A] [answered]
<img width="955" alt="image" src="https://github.com/wafflestudio/seminar-2023/assets/76721027/56f7819f-c945-466c-b46e-c26e996db751">
위는 테스트케이스 성공한 JPQL입니다. 그 과정 속에서 몇가지 질문이 있어 올립니다.

1. Query 어노테이션을 사용하게 되면 해당 쿼리가 실행되는데 그렇다면 함수명에 따라 작동하던 JPA는 어떻게 되는 것인가요? 함수명도 쿼리에 맞게 일치시켜주는 것이 올바를까요?
2. JPQL을 이용하게 되면 쿼리문이 긴 경우에 다소 가독성이 떨어진다는 생각이 들었습니다. 이러한 부분은 성능을 위해 어쩔 수 없는 것인가요?
3. Query 어노테이션을 사용하는 경우 Eager 타입인 필드(Song 엔티티의 경우 Album 엔티티)도 직접 join fetch를 해주어야 가져오는 것을 확인했습니다. 즉 이는 쿼리 어노테이션을 사용하게 될 경우 해당 쿼리가 우선시되어 기존에 가져오던 Eager 타입들은 가져오지 않게 되는 것인가요?
<img width="841" alt="image" src="https://github.com/wafflestudio/seminar-2023/assets/76721027/09f2e49c-5102-469f-8afd-917ff9abcff3">

![image](https://github.com/wafflestudio/seminar-2023/assets/76721027/f4521cba-4ec4-4488-b9fa-d4fabb5d049e)
위처럼 album을 join fetch 하지 않았더니 처음 쿼리 때 가져오지 않고 매번 쿼리를 날렸습니다.

4. 위처럼 쿼리를 날릴 경우에는 앨범에 대한 모든 데이터, 아티스트에 대한 모든 데이터를 가져오는데 막상 원하는 열들만 select을 하면 오류가 났습니다. 이러한 일이 생기는 이유도 궁금합니다.
<img width="1135" alt="image" src="https://github.com/wafflestudio/seminar-2023/assets/76721027/797b2939-df21-4ac7-b662-8562f7d00ad7">

![image](https://github.com/wafflestudio/seminar-2023/assets/76721027/81f93a9f-4271-44dd-819d-b1f02a2305e6)
이렇게 여러개의 중복된 엔티티가 생기더군요

하다보니 질문이 좀 많이 생겨서..
항상 감사합니다(__)

**Comment by PFCJeong** [comment, marked as answer]
안녕하세요~ 좋은 질문 많이 남겨주셨네요.

1. JPQL이 우선적으로 적용되는 것 같습니다. JPQL을 사용했음을 명시하도록 함수명을 짓는 것이 좋겠습니다.
2. 아무래도 쿼리를 직접 작성하다보니 실수의 여지도 있고 가독성도 떨어지긴 합니다. 아래와 같이 적절히 개행을 해서 가독성을 조금이라도 높일 수도 있는데요. 세미나에서 다루지는 않았지만 QueryDSL을 쓰는 방법도 있습니다. 잘못된 쿼리는 성능에 큰 영향을 미치기 때문에 가독성보다는 성능을 신경쓰는게 맞을 것 같습니다.
```kotlin
@Query("""
    SELECT s FROM songs s
    JOIN FETCH s.album al 
    JOIN FETCH s.artists sa
    JOIN FETCH sa.artist
    WHERE s.title LIKE %:keyword%
""")
```
3. 네 맞습니다. 프로퍼티에 명시한 FetchType.EAGER보다 JPQL로 작성한 쿼리가 우선시 됩니다.
4. 특정 칼럼만 가져오고 싶으면 Interface Projection을 사용해야합니다. 위와 같이 JQPL을 작성해도, `a.image, ar.name`만 조회하는 것이 아니고 album과 artist의 모든 칼럼들을 전부 다 조회합니다. 
albums.image만 조회해왔다고 했을 때, `song(SongEntity).album(AlbumEntity).title`는 비어있게 되기 때문입니다. 
흠,, 그런데 왜 중복된 엔티티를 리턴하게 되는지 저도 잘 모르겠습니다. (song 1538의 artist가 3명인 것이 연관되어 있을 것 같네요)

감사합니다.

**Reply by DoohyunHwang97** [reply]
모든 질문에 명쾌하게 해답을 주시다니 감사합니다 :)
